src/rules/maxNestingDepthRule.ts

```diff
@@ -0,0 +1,142 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isFunctionScopeBoundary, isIdentifier } from "tsutils";
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static DEFAULT_THRESHOLD = 4;
+    public static MINIMUM_THRESHOLD = 0;
+
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "max-nesting-depth",
+        description: "Enforces a maximum threshold of nesting depth.",
+        descriptionDetails: Lint.Utils.dedent`
+            The nesting depth of the code blocks is a metric determined for each method.
+            It is computed by determining the nesting levels of the control statements within the method.
+            A starting value of 0 is assigned, then this value is
+            incremented each time a new block is entered, and decremented
+            each time a block is left.
+            The following statements contribute to nesting depth:
+            * \`if\` statements
+            * \`switch\` statements
+            * \`for\`, \`for in\` and \`for of\` loops
+            * \`while\` and \`do while\` loops
+            * \`try\` \`catch\` and \`finally\` blocks`,
+        rationale: Lint.Utils.dedent`
+            Maximum nesting depth is a code metric which indicates the level of imbrication in a
+            function. This metric quickly reveals the methods whose code is made difficult to read
+            by an excessive nesting of the control statements.
+
+            It's better to have smaller, single-purpose functions with self-documenting names.`,
+        optionsDescription: Lint.Utils.dedent`
+            An optional upper limit for maximum nesting depth can be specified. If no limit option
+            is provided a default value of ${Rule.DEFAULT_THRESHOLD} will be used.`,
+        options: {
+            type: "number",
+            minimum: Rule.MINIMUM_THRESHOLD,
+        },
+        optionExamples: [true, [true, 5]],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(expected: number, actual: number, name?: string): string {
+        return (
+            `The function${name === "" ? "" : ` ${name}`} has a nesting depth of at least ` +
+            `${actual} which is higher than the threshold of ${expected}`
+        );
+    }
+
+    public isEnabled(): boolean {
+        // Disable the rule if the option is provided but non-numeric or less than the minimum.
+        const isThresholdValid =
+            typeof this.threshold === "number" && this.threshold >= Rule.MINIMUM_THRESHOLD;
+        return super.isEnabled() && isThresholdValid;
+    }
+
+    public get threshold(): number {
```

Comment:
> public

Nit: can this be `private` please? Better to indicate that `threshold` isn't used anywhere else.

Reply:
Of course !